### ISSUE TITLE
JSON Path parsing issue PO040

### ISSUE THREAD

**Issue by PratMoha** [apigee/apigeelint#562]
Hi @DinoChiesa ,

I'm currently facing an issue in my deployment pipeline where the apigee linter is failing with below error:

```
sharedflowbundle/policies/EV-GetRichDetailsJsonOutput.xml(7,13): error PO040 : JSONPath ({flow.error.extension.rich-details-json-path}) is invalid (Error: Lexical error on line 1. Unrecognized text.
{flow.error.extensio
^)
```

I'm dynamically assigning the JSON path to be extracted by my error handler. The value passed to the variable are: **$.msg, $.message, $.error**
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ExtractVariables async="false" continueOnError="true" enabled="true" name="EV-GetErrorDetails">
    <DisplayName>EV-GetErrorDetails</DisplayName>
    <Properties/>
    <JSONPayload>
        <Variable name="flow.error.extension.rich-details-json-output">
            **<JSONPath>{flow.error.extension.rich-details-json-path}</JSONPath>**
        </Variable>
    </JSONPayload>
    <Source clearPayload="false">flow.error.extension.rich-details-json-input</Source>
    <IgnoreUnresolvedVariables>true</IgnoreUnresolvedVariables>
</ExtractVariables>
```

It seems the lint is failing during the jsonpath parse. 

Regards,
Pratyush

**Comment by DinoChiesa** [comment]
Despite the documentation that shows {example} in the place where the JSONPath should be, the string within JSONPath is a static string.  In other words,  in the current policy, this notation: 
```xml
    <JSONPayload>
        <Variable name="flow.error.extension.rich-details-json-output">
            <JSONPath>{flow.error.extension.rich-details-json-path}</JSONPath>
        </Variable>
    </JSONPayload>
```

...is ineffective. At runtime the policy will not de-reference the thing within curly braces.  The policy will treat `{flow.error.extension.rich-details-json-path}` as a JSONPath, and that string will fail compilation as a JSONPath. The Apigeelint plugin is trying to do the same thing - compile the expression - and is returning the error it encounters. 

So I think it's correct behavior on the part of the plugin here. 

You can do this: 
```xml
    <JSONPayload>
        <Variable name="flow.error.extension.rich-details-json-output">
            <JSONPath>$.some.path[1]</JSONPath>
            <JSONPath>$.another.path</JSONPath>
        </Variable>
    </JSONPayload>
```

In other words, multiple JSONPath elements. But each one must be actual JSONPath, directly specified. 

Stepping back from the plugin or Apigeelint, but . . . . If you want to evaluate a "dynamic" JSONPath within Apigee, and store the result into a variable, then you need to use the AssignMessage / AssignVariable with the message template, using a jsonPath static function, which takes a variable. like this:

```xml
<AssignMessage name='AM-Set-Var-from-Dynamic-JsonPath'>
  <AssignVariable>
    <Name>extracted_element</Name>
    <Template>{jsonPath(var-with-jsonpath-in-it,var-containing-json)}</Template>
  </AssignVariable>
</AssignMessage>
```

**Comment by DinoChiesa** [comment]
YES

I tested as well after making that comment.  My prior comment is not correct.

**Comment by PratMoha** [comment]
Hi Dino/Sai

Thank you for acknowledging and the same is stated in the documentation : https://cloud.google.com/apigee/docs/api-platform/reference/policies/extract-variables-policy

Note: The <JSONPath> element supports the dynamic string substitution feature called message templating

Regards,
Pratyush

**Comment by DinoChiesa** [comment]
Yes, you are correct Pratyush.
